### PR TITLE
Hash submitters in the deduplication key [DPP-347]

### PR DESCRIPTION
**Pull request by kamil-da**
## Motivation
Submitting a command with a large number of submitters results in:
```
org.postgresql.util.PSQLException: ERROR: index row size 6512 exceeds maximum 2712 for index "participant_command_submissions_pkey"
  Hint: Values larger than 1/3 of a buffer page cannot be indexed.
Consider a function index of an MD5 hash of the value, or use full text indexing.
```
due to how we construct deduplication keys. Concatenating a large number of submitters causes exceeding Postgres limits.

## Main changes
Made the deduplication key containing a hash of submitters instead of a concatenated string of submitters.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+      request = ledger
+        .submitRequest(
+          actAs = parties.toList,
+          readAs = parties.toList,
+          commands = DummyWithAnnotation(parties.head, "First submission").create.command,
+        )
+        .update(
+          _.commands.deduplicationTime := deduplicationTime.asProtobuf
+        )
+      _ <- ledger.submit(request)
```
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:44]: Shouldn't this test assert that the dummy contract was successfully created?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import scala.concurrent.Future
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+import scala.util.Random
+
+final class ValueLimitsIT(timeoutScaleFactor: Double) extends LedgerTestSuite {
+
+  /** Postgres has a limit on the index row size of 2712.
+    * We test that a large submitters number doesn't cause deduplication failure because of that limit.
+    * THIS TEST CASE DOES NOT TEST ANY OTHER FAILURES THAT MAY BE CAUSED BY A LARGE SUBMITTERS NUMBER
+    */
```
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:20]: This description is not accurate:
- The test uses the ledger API and not the internals of the indexer,  so you don't really know what kind of failures it tests for.
- The description mentions a limit of Postgres, but you don't know whether the test runs against a Postgres index database.
- You are not checking any property of command deduplication, you are just submitting a request with a large number of parties. I would name the test `"Submit commands with a large number of submitters"`.

**Review by rautenrieth-da — CHANGES_REQUESTED** [review]
Any change in `com.daml.platform.store.dao` needs to be mirrored in `com.daml.platform.store.appendonlydao`, can you please edit `JdbcLedgerDao` there accordingly?

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala (hunk 1 above)
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:42]: Fixed 👍

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala (hunk 2 above)
Comment (kamil-da) [reply ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala]: @rautenrieth-da Thank you for the review!
I've modified the test suite according to your suggestions. I've also added a change to a database index that was blocking submissions due to index row size limits.

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        .submitRequest(
+          actAs = parties.toList,
+          readAs = parties.toList,
+          commands = DummyWithAnnotation(parties.head, "First submission").create.command,
+        )
+        .update(
+          _.commands.deduplicationTime := deduplicationTime.asProtobuf
+        )
+      _ <- ledger.submit(request)
+      _ <- Delayed.by(1.second)(())
```
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:43]:
```suggestion
      _ <- ledger.submitAndWait(request)
```

File: ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        )
+      }
+      request = ledger
+        .submitRequest(
+          actAs = parties.toList,
+          readAs = parties.toList,
+          commands = DummyWithAnnotation(parties.head, "First submission").create.command,
+        )
+        .update(
+          _.commands.deduplicationTime := deduplicationTime.asProtobuf
```
Comment (rautenrieth-da) [inline ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/suites/ValueLimitsIT.scala:40]: Why do you need to set a custom deduplication time?

**Review by rautenrieth-da — CHANGES_REQUESTED** [review]
Left some minor comments. You still need to do the change in `com.daml.platform.store.appendonlydao` though.

**Comment by kamil-da** [comment]
@rautenrieth-da I've adapted the PR according to your review and the discussion during the meeting.
- implemented the feature for `appendonlydao` package
- removed changes to the database index - will come in a follow-up PR
- removed `ValueLimitsIT` suite - will come in a follow-up PR

For this PR I've replaced API-level testing with simple unit testing because it's currently impossible to make API-level tests pass without making changes to database indexes.

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/dao/DeduplicationKeyMakerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+  val commandId: CommandId = CommandId(Ref.LedgerString.assertFromString(UUID.randomUUID.toString))
+
+  "DeduplicationKeyMaker" should {
+    "make a deduplication key starting with a command ID in plain-text" in {
+      DeduplicationKeyMaker.make(commandId, List(aParty())) should startWith(commandId.unwrap)
+    }
+
+    "make a deduplication key with a limited length for a large number of submitters" in {
+      val submitters = (1 to 50).map(_ => aParty()).toList
+      val MaxKeyLength = 200
```
Comment (rautenrieth-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/store/dao/DeduplicationKeyMakerSpec.scala:26]: This constant seems arbitrary, but the value is fine. It's smaller than the maximum length for indexed string columns in both Postgres and Oracle (which was the reason for this change). Maybe add a comment?

**Review by rautenrieth-da — APPROVED** [review]
Thank you for deduplicating the deduplication logic 🙂

File: ledger/participant-integration-api/src/test/suite/scala/platform/store/dao/DeduplicationKeyMakerSpec.scala (hunk 5 above)
Comment (kamil-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/store/dao/DeduplicationKeyMakerSpec.scala:26]: Good point - I'll comment on motivation for this limit.

**Pull request merged**